pkg/compact/compact.go

```diff
@@ -200,131 +163,109 @@ func (c *Syncer) syncMetas(ctx context.Context) error {
-		// This prevents re-compacting the same data. The block remain in the top-level
-		// directory until it has actually been deleted from the bucket.
-		if err := c.add(dir); err != nil {
-			level.Warn(c.logger).Log("msg", "add new block to group", "err", err)
-		}
+		remote[id] = struct{}{}
+		c.blocks[id] = &meta
+
 		return nil
 	})
 	if err != nil {
 		return errors.Wrap(err, "retrieve bucket block metas")
 	}
 
 	// Delete all local block dirs that no longer exist in the bucket.
-	for id, meta := range local {
-		if _, ok := remote[id]; ok {
-			continue
-		}
-
-		if err := os.RemoveAll(filepath.Join(c.dir, id.String())); err != nil {
-			level.Warn(c.logger).Log("msg", "delete outdated block", "block", id, "err", err)
-		}
-		if err := os.RemoveAll(filepath.Join(c.dir, "groups", c.groupKey(meta), id.String())); err != nil {
-			level.Warn(c.logger).Log("msg", "delete outdated block from group", "block", id, "err", err)
+	for id := range c.blocks {
+		if _, ok := remote[id]; !ok {
+			delete(c.blocks, id)
 		}
 	}
-
 	return nil
 }
 
-func (c *Syncer) groupKey(meta *block.Meta) string {
-	return fmt.Sprintf("%x", labels.FromMap(meta.Thanos.Labels).Hash())
+// groupKey returns a unique identifier for the group the block belongs to. It considers
+// the downsampling resolution and the block's labels.
+func groupKey(meta *block.Meta) string {
+	return fmt.Sprintf("%d@%s", meta.Thanos.Downsample.Resolution, labels.FromMap(meta.Thanos.Labels))
 }
 
-// add adds the block in the given directory to its respective compaction group.
-func (c *Syncer) add(bdir string) error {
-	meta, err := block.ReadMetaFile(bdir)
-	if err != nil {
-		return errors.Wrap(err, "read meta file")
-	}
-	h := c.groupKey(meta)
-
+// Groups returns the compaction groups for all blocks currently known to the syncer.
+func (c *Syncer) Groups() (res []*Group, err error) {
 	c.mtx.Lock()
 	defer c.mtx.Unlock()
 
-	g, ok := c.groups[h]
-	if !ok {
-		g, err = newGroup(
-			log.With(c.logger, "compactionGroup", h),
-			c.reg,
-			c.bkt,
-			filepath.Join(c.dir, "groups", h),
-			meta.Thanos.Labels,
-		)
-		if err != nil {
-			return errors.Wrap(err, "create compaction group")
+	groups := map[string]*Group{}
+
+	for _, m := range c.blocks {
+		g, ok := groups[groupKey(m)]
+		if !ok {
+			g, err = newGroup(
+				log.With(c.logger, "compactionGroup", groupKey(m)),
+				c.reg,
+				c.bkt,
+				labels.FromMap(m.Thanos.Labels),
+				m.Thanos.Downsample.Resolution,
+			)
+			if err != nil {
+				return nil, errors.Wrap(err, "create compaction group")
+			}
+			groups[groupKey(m)] = g
+			res = append(res, g)
 		}
-		c.groups[h] = g
-	}
-	return g.Add(meta)
-}
-
-// Groups returns the compaction groups created by the Syncer.
-func (c *Syncer) Groups() (res []*Group) {
-	c.mtx.Lock()
-	defer c.mtx.Unlock()
-
-	for _, g := range c.groups {
-		res = append(res, g)
+		g.Add(m)
 	}
-	return res
+	sort.Slice(res, func(i, j int) bool {
+		return res[i].Key() < res[j].Key()
```

Comment:
Do we really want this order?
```
1m@block1
1m@block2
1m@block3
5m@block1
5m@block2
...
```
Maybe we should put labels in  the front of  `groupKey`?  Then we group by labels first.

I don't know which one is better ;p Just checking if there are any rationales behind that order (:

Reply:
Mh, no strict rationale I guess. But it likely makes sense to treat lower resolution blocks with higher priority since the most recent ones are queried the most.

Reply:
cool